cloud/scope/vpc_cluster.go

```diff
@@ -686,3 +742,321 @@ func (s *VPCClusterScope) buildCOSObjectHRef() (*string, error) {
+	if s.IBMVPCCluster.Spec.Network.ControlPlaneSubnets == nil || len(s.IBMVPCCluster.Spec.Network.ControlPlaneSubnets) == 0 {
+		subnets, err = s.buildSubnetsForZones()
+		if err != nil {
+			return false, fmt.Errorf("error failed building control plane subnets: %w", err)
+		}
+	} else {
+		subnets = s.IBMVPCCluster.Spec.Network.ControlPlaneSubnets
+	}
+
+	// Reconcile Control Plane subnets.
+	requeue := false
+	for _, subnet := range subnets {
+		if requiresRequeue, err := s.reconcileSubnet(subnet, true); err != nil {
+			return false, fmt.Errorf("error failed reconciling control plane subnet: %w", err)
+		} else if requiresRequeue {
+			// If the reconcile of the subnet requires further reconciliation, plan to requeue entire ReconcileSubnets call, but attempt to further reconcile additional Subnets (attempt all subnet reconciliation).
+			requeue = true
+		}
+	}
+
+	// If no Worker subnets were supplied, attempt to create one in each zone.
+	if s.IBMVPCCluster.Spec.Network.WorkerSubnets == nil || len(s.IBMVPCCluster.Spec.Network.WorkerSubnets) == 0 {
+		// If neither Control Plane nor Worker subnets were supplied, we rely on both Planes using the same subnet per zone, and we will re-reconcile those subnets below, for IBMVPCCluster Status updates.
+		if len(s.IBMVPCCluster.Spec.Network.ControlPlaneSubnets) != 0 {
+			subnets, err = s.buildSubnetsForZones()
+			if err != nil {
+				return false, fmt.Errorf("error failed building worker subnets: %w", err)
+			}
+		}
+	} else {
+		subnets = s.IBMVPCCluster.Spec.Network.WorkerSubnets
+	}
+
+	// Reconcile Worker subnets.
+	for _, subnet := range subnets {
+		if requiresRequeue, err := s.reconcileSubnet(subnet, false); err != nil {
+			return false, fmt.Errorf("error failed reconciling worker subnet: %w", err)
+		} else if requiresRequeue {
+			// If the reconcile of the subnet requires further reconciliation, plan to requeue entire ReconcileSubnets call, but attempt to further reconcile additional Subnets (attempt all subnet reconciliation).
+			requeue = true
+		}
+	}
+
+	// Return whether or not one or more subnets required further reconciling after attempting to process all Control Plane and Worker subnets.
+	return requeue, nil
+}
+
+// reconcileSubnet will attempt to find the existing subnet, or create it if necessary.
+// The logic can handle either Control Plane or Worker subnets, but must distinguish between them for Status updates.
+func (s *VPCClusterScope) reconcileSubnet(subnet infrav1beta2.Subnet, isControlPlane bool) (bool, error) { //nolint: gocyclo
+	// If subnet already has a Name defined, use that for lookup.
+	if subnet.Name != nil {
+		// If we have the subnet name, we can easily check Network Stauts on the subnet's status.
+		if isControlPlane {
+			// If we find the subnet in Control Plane subnet status and that it is ready, we can return, with no requeue required.
+			if subnet, ok := s.NetworkStatus().ControlPlaneSubnets[*subnet.Name]; ok && subnet.Ready {
```

Comment:
Even if we find it in status, don't we need to check the cloud that its still in desired state?
Please take care of this in all the similar blocks.

Reply:
The discussion I had previously, was that with certain VPC resources, if the details are already in Status, checking the realtime status of said resource would not need to happen. For instance, handling cases where CAPI created a resource, and someone/thing deleted the resource, having CAPI attempt to re-reconcile that case, even though Status had a name and ID (which are no longer valid, like ID) would not be desirable. 

As VPC and Subnets, do not typically change status, unlike Load Balancers, the expectation is that once these resources are Ready (or the equivalent VPC status value/constant) it should stay that way until deletion (either requested or done maliciously). Load Balancers, due to `updating` being a frequent occurrence, likely would not be able to rely on previous status (instead checking status in realtime).

If this expectation has changed, and CAPI is expected to recreate all VPC resources (if they disappear, API fails to find them, etc.), I can change all resource logic to no rely on Status details (such as ID) if that changes in realtime.

Reply:
@Karthik-K-N @mkumatag could you please share your opinion here?

Reply:
We usually follow this pattern of fetching id from status and checking its actual status in cloud, Its to make sure that the observed resource status is correct and up to date with real world.

I think its better to follow this pattern as its helps in early detection of any potential problems.

Reply:
Should reconciliation completely stop (panic) if a resource cannot be found then?
Or should resources be recreated and update Status with new ID's, etc.?

Reply:
As of now reconciler will throw error. For [reference](https://github.com/Karthik-K-N/cluster-api-provider-ibmcloud/blob/2c25714f46ca572018419ebb3cb5ad6fe93793ca/cloud/scope/powervs_cluster.go#L994-L1016). Its upto the user to resolve the error.

Reply:
I will attempt to refactor, with the expectation that anything in Status that cannot be found in IBM Cloud, etc. results in an continuous error. Other resources will be updated in the same fashion in other PR's then.